### PR TITLE
Changes to Controller implementation

### PR DESCRIPTION
**Pull request by snaselj**
# Closes Nan

## What's Changed

- Added docs and ER diagram https://github.com/nautobot/nautobot/blob/u/snaselj-5325-wip/nautobot/docs/development/core/controllers.md
- Added fixtures for testing https://github.com/nautobot/nautobot/blob/u/snaselj-5325-wip/nautobot/dcim/fixtures/controller-data.json
- Added comments to models.
- Fixed `ControllerDeviceGroup.controller.related_name` to controller_device_groups as I don't see any reason not to publish this in `Controller` views.
- Implemented API and UI classes.

### CODE REVIEW

**Comment by snaselj** [comment]
WIP, however interested in some feedback @glennmatthews

File: nautobot/docs/development/core/controllers.md [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+A software version for a Device, Virtual Machine or Inventory Item.
+
+### `Device`
+
+A Device represents a piece of physical hardware. Each Device is assigned a `DeviceType`,
+`Role`, and (optionally) a `Platform`. Device names are not required, however if one is set it must be unique.
+
+When a new `Device` is created, console/power/interface/device bay components are created along with it as dictated
+by the component templates assigned to its `DeviceType`. Components can also be added, modified, or deleted after the
+creation of a Device.
```
Comment (glennmatthews) [inline nautobot/docs/development/core/controllers.md:58]: Seems like unnecessary/redundant information in this context? Might be better to instead discuss how a Device relates to a Controller and/or ControllerDeviceGroup?

File: nautobot/dcim/factory.py [hunk 2]
```diff
@@ -669,3 +671,35 @@ class Params:
     documentation_url = factory.Maybe("has_documentation_url", factory.Faker("uri"), "")
     long_term_support = NautobotBoolIterator()
     pre_release = NautobotBoolIterator()
+
+
+class ControllerFactory(PrimaryModelFactory):
```
[the remark this answers is not in the document]
Comment (snaselj) [reply nautobot/dcim/factory.py:676]: Keeping for now as is, will resolve this with tests.

File: nautobot/dcim/factory.py [hunk 3]
```diff
@@ -669,3 +671,35 @@ class Params:
...
+    name = factory.Faker("word")
+    description = factory.Faker("sentence")
+    status = random_instance(lambda: Status.objects.get_for_model(Controller), allow_null=False)
+    role = random_instance(lambda: Role.objects.get_for_model(Controller), allow_null=False)
+    platform = random_instance(Platform, allow_null=False)
+    location = random_instance(Location, allow_null=False)
+    tenant = random_instance(Tenant, allow_null=True)
+    external_integration = random_instance(ExternalIntegration, allow_null=True)
+    deployed_controller_device = factory.Maybe("has_device", random_instance(Device), None)
+    # deployed_controller_group = factory.Maybe("has_device", random_instance(DeviceRedundancyGroup), None)
```
[the remark this answers is not in the document]
Comment (snaselj) [reply nautobot/dcim/factory.py:692]: Correct, keeping for now as is, will resolve this with tests.

File: nautobot/dcim/filters/__init__.py [hunk 4]
```diff
@@ -1826,3 +1830,79 @@ class DeviceTypeToSoftwareImageFileFilterSet(BaseFilterSet):
...
+        filter_predicates={
+            "name": "icontains",
+        }
+    )
+    controller = NaturalKeyOrPKMultipleChoiceFilter(
+        queryset=Controller.objects.all(),
+        to_field_name="name",
+        label="Controller (name or ID)",
+    )
+    parent = NaturalKeyOrPKMultipleChoiceFilter(
```
[the remark this answers is not in the document]
Comment (snaselj) [reply nautobot/dcim/filters/__init__.py:1890]: `subtree` added.

File: nautobot/dcim/forms.py [hunk 5]
```diff
@@ -4091,3 +4097,122 @@ class SoftwareVersionForm(NautobotModelForm):
...
         model = SoftwareVersion
         fields = "__all__"
+
+
+class ControllerForm(LocatableModelFormMixin, NautobotModelForm, TenancyForm, LocalContextModelForm):
+    """Controller credit/edit form."""
+
+    class Meta:
+        model = Controller
+        fields = "__all__"
```
[the remark this answers is not in the document]
Comment (snaselj) [reply nautobot/dcim/forms.py:4101]: IIRC, displayed fields are managed in the `retrieve` form template file for the detail form.

File: nautobot/dcim/api/views.py [hunk 6]
```diff
@@ -818,3 +820,20 @@ class DeviceTypeToSoftwareImageFileViewSet(ModelViewSet):
...
     filterset_class = filters.DeviceTypeToSoftwareImageFileFilterSet
+
+
+#
+# Controllers
+#
+
+
+class ControllerViewSet(NautobotModelViewSet):
+    queryset = Controller.objects.select_related("location", "platform", "role", "tenant", "status")
```
[the remark this answers is not in the document]
Comment (snaselj) [reply nautobot/dcim/api/views.py]: Added

File: nautobot/dcim/filters/__init__.py [hunk 7]
```diff
@@ -1826,3 +1830,79 @@ class DeviceTypeToSoftwareImageFileFilterSet(BaseFilterSet):
...
+        filter_predicates={
+            "name": "icontains",
+            "description": "icontains",
+        }
+    )
+    location = NaturalKeyOrPKMultipleChoiceFilter(
+        queryset=Location.objects.all(),
+        to_field_name="name",
+        label="Location (name or ID)",
+    )
```
[the remark this answers is not in the document]
Comment (snaselj) [reply nautobot/dcim/filters/__init__.py]: Removed

File: nautobot/dcim/filters/__init__.py [hunk 8]
```diff
@@ -1826,3 +1830,79 @@ class DeviceTypeToSoftwareImageFileFilterSet(BaseFilterSet):
...
+    platform = NaturalKeyOrPKMultipleChoiceFilter(
+        queryset=Platform.objects.all(),
+        to_field_name="name",
+        label="Platform (name or ID)",
+    )
+    tenant = NaturalKeyOrPKMultipleChoiceFilter(
+        queryset=Tenant.objects.all(),
+        to_field_name="name",
+        label="Tenant (name or ID)",
+    )
```
[the remark this answers is not in the document]
Comment (snaselj) [reply nautobot/dcim/filters/__init__.py]: Removed

File: nautobot/dcim/forms.py [hunk 9]
```diff
@@ -4091,3 +4097,122 @@ class SoftwareVersionForm(NautobotModelForm):
...
+    """Controller basic filter form."""
+    model = Controller
+    q = forms.CharField(required=False, label="Search")
+    name = forms.CharField(required=False, label="Name")
+    description = forms.CharField(required=False, label="Description")
+    location = DynamicModelMultipleChoiceField(
+        queryset=Location.objects.all(),
+        required=False,
+        label="Location",
+    )
```
[the remark this answers is not in the document]
Comment (snaselj) [reply nautobot/dcim/forms.py]: Removed

File: nautobot/dcim/forms.py [hunk 10]
```diff
@@ -4091,3 +4097,122 @@ class SoftwareVersionForm(NautobotModelForm):
...
+    platform = DynamicModelMultipleChoiceField(
+        queryset=Platform.objects.all(),
+        required=False,
+        label="Platform",
+    )
+    tenant = DynamicModelMultipleChoiceField(
+        queryset=Tenant.objects.all(),
+        required=False,
+        label="Tenant",
+    )
```
[the remark this answers is not in the document]
Comment (snaselj) [reply nautobot/dcim/forms.py]: Removed

File: nautobot/dcim/forms.py [hunk 11]
```diff
@@ -4091,3 +4097,122 @@ class SoftwareVersionForm(NautobotModelForm):
...
+    StatusModelBulkEditFormMixin,
+    RoleModelBulkEditFormMixin,
+    NautobotBulkEditForm,
+    LocalContextModelBulkEditForm,
+):
+    """Controller bulk edit form."""
+
+    class Meta:
+        model = Controller
+        fields = "__all__"
```
[the remark this answers is not in the document]
Comment (snaselj) [reply nautobot/dcim/forms.py]: IIRC, displayed fields are managed in the `retrieve` form template file for the detail form.

Updated fields on the rest of forms, where html templates are not used.

File: nautobot/dcim/models/devices.py [hunk 12]
```diff
@@ -1293,26 +1299,51 @@ class Controller(PrimaryModel, ConfigContextModel):
...
+        ordering = ("_name",)
+
+    def __str__(self):
+        return self.display or super().__str__()
+
+    @property
+    def display(self):
+        if self.name:
+            return self.name
+        return f"{self.platform} - {self.role}"
```
[the remark this answers is not in the document]
Comment (snaselj) [reply nautobot/dcim/models/devices.py]: Simplified

File: nautobot/dcim/models/devices.py [hunk 13]
```diff
@@ -1293,26 +1299,51 @@ class Controller(PrimaryModel, ConfigContextModel):
...
-    weight = models.PositiveIntegerField(default=1000)
+    This model allows for the organization of controlled devices into hierarchical groups for structured representation.
+    """
+
+    name = models.CharField(
+        max_length=CHARFIELD_MAX_LENGTH,
+        unique=True,
+        help_text="Name of the controller device group",
+    )
+    _name = NaturalOrderingField(target_field="name", max_length=CHARFIELD_MAX_LENGTH, db_index=True)
```
[the remark this answers is not in the document]
Comment (snaselj) [reply nautobot/dcim/models/devices.py]: Removed

File: nautobot/dcim/models/devices.py [hunk 14]
```diff
@@ -1293,26 +1299,51 @@ class Controller(PrimaryModel, ConfigContextModel):
...
         on_delete=models.CASCADE,
-        related_name="+",
+        related_name="controller_device_groups",
         blank=False,
         null=False,
+        help_text="Controller that manages the devices in this group",
     )
+
+    class Meta:
+        ordering = ("_name",)
```
[the remark this answers is not in the document]
Comment (snaselj) [reply nautobot/dcim/models/devices.py]: Inserted `weight` field before `name`.

**Comment by snaselj** [comment]
All comments are addressed, fixed all linter issues.

File: nautobot/dcim/forms.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (glennmatthews) [reply nautobot/dcim/forms.py:4101]: Should only be the case if you're writing a custom edit form - in the default template case, field_order should be respected AFAIK.

[the remark this answers is not in the document]
Comment (snaselj) [reply nautobot/dcim/forms.py:4101]: The form was not created properly without the template. Only `Tags` section was on `Controller` tab, without any other fields regardless of `Meta.fields` content, tried various combinations.

[the remark this answers is not in the document]
Comment (glennmatthews) [reply nautobot/dcim/forms.py:4101]: To be clear, I'm talking about the create/edit form, not the retrieve/detail template.

File: nautobot/dcim/models/devices.py [hunk 15]
```diff
@@ -1293,26 +1298,51 @@ class Controller(PrimaryModel, ConfigContextModel):
         null=True,
     )
 
+    class Meta:
+        ordering = ("name",)
+
+    def __str__(self):
+        return self.display or super().__str__()
```
Comment (whitej6) [inline nautobot/dcim/models/devices.py:1305]: IIRC the display attr is a property that is generated using the `__str__`. You should not create the display property

**Review by glennmatthews — APPROVED** [review]
Looks like good incremental progress to me.

File: nautobot/dcim/forms.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (snaselj) [reply nautobot/dcim/forms.py]: Sorry, my bad, fixed.

**Pull request merged**